### PR TITLE
Convert the Convenience Inits to designated Inits to use them for subclassing.

### PR DESCRIPTION
**Pull request by Narsail**
I tried to subclass the CollectionProvider today and got an error that I can't use the convenience Inits (like used in the documentation) for subclassing.

My proposal is to convert the convenience inits to designated inits. Sadly simply removing the convenience mark resulted in an error:
````
Designated initializer for cannot ... delegate (with 'self.init')
````

I know that simply duplicating code isn't really helpful so if someone has a better solution to make those Inits subclassable, you'r welcome :)

### CODE REVIEW

**Comment by jindulys** [comment]
You don't need to change `CollectionProvider`'s convenient initializers to a designated one.

What you can do is provide a subclass that override `CollectionProvider`'s designated one and then you can get `CollectionProvider`'s convenient initializers for free.

```Swift
class MyProvider<Data, View: UIView>: CollectionProvider<Data, View> {
    override init(identifier: String? = nil,
                dataProvider: DataProvider,
                viewProvider: ViewProvider,
                layout: Layout = FlowLayout<Data>(),
                sizeProvider: @escaping SizeProvider = defaultSizeProvider,
                presenter: Presenter? = nil,
                willReloadHandler: (() -> Void)? = nil,
                didReloadHandler: (() -> Void)? = nil,
                tapHandler: TapHandler? = nil) {
        super.init(identifier: identifier,
                   dataProvider: dataProvider,
                   viewProvider: viewProvider,
                   layout: layout,
                   sizeProvider: sizeProvider,
                   presenter:presenter,
                   willReloadHandler: willReloadHandler,
                   didReloadHandler: didReloadHandler,
                   tapHandler: tapHandler)
    }
}
```

Then you can use `MyProvider` in your view controller.

```Swift
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let provider = MyProvider(
            data: testData,
            viewUpdater: { (view: ContentView, data: OffsetData, at: Int) in
                view.populate(data: data)
            }
        )
        provider.layout = OffsetLayout()
        provider.sizeProvider = OffsetDataSizeProvider
        collectionView.provider = provider
    }
```

**Comment by Narsail** [comment]
I don’t want to get them for free. I want to use the convenience inits within my own simple init function. But you can’t call a convenience init within a designated custom init of the subclass.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lkzhao** [comment]
I think this kinda make sense. I also felt the need to call one of these convenient initializer in one of my subclass. @jindulys should we consider merging this in?

File: Sources/Provider/CollectionProvider.swift [hunk 1]
```diff
@@ -50,46 +50,48 @@ open class CollectionProvider<Data, View: UIView>: BaseCollectionProvider {
...
+              dataProvider: DataProvider,
+              viewGenerator: ((Data, Int) -> View)? = nil,
+              viewUpdater: @escaping (View, Data, Int) -> Void,
+              layout: Layout = FlowLayout<Data>(),
+              sizeProvider: @escaping SizeProvider = defaultSizeProvider,
+              presenter: Presenter? = nil,
+              willReloadHandler: (() -> Void)? = nil,
+              didReloadHandler: (() -> Void)? = nil,
+              tapHandler: TapHandler? = nil) {
+
```
Comment (jindulys) [inline Sources/Provider/CollectionProvider.swift]: Extra blank line

**Comment by Narsail** [comment]
Sry i was on holiday and wasn‘t able to correct the code. Will happen in the next days.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Pull request merged**